src/types.ts

```diff
@@ -1,79 +1,195 @@
+    /** DEPRECATED */
     getSettings?: (meta: Meta<Input>) => PluginSettings
-    /** Receive a single non-snapshot event and return it in its processed form. You can discard the event by returning null. */
+    /** Used for filtering events, return True to keep the event and False to drop it. */
+    filterEvent?: (event: PostHogEvent, meta: Meta<Input>) => boolean
+    /** Used for modifying events, return the updated event */
+    modifyEvent?: (event: PostHogEvent, meta: Meta<Input>) => PostHogEvent
+    /** DEPRECATED: use filterEvent or modifyEvent */
     processEvent?: (event: PluginEvent, meta: Meta<Input>) => PluginEvent | null | Promise<PluginEvent | null>
-    /** DEPRECATED: Receive a batch of events and return it in its processed form. You can discard events by not including them in the returned array. You can also append additional events to the returned array. */
+    /** DEPRECATED: use filterEvent or modifyEvent */ 
     processEventBatch?: (eventBatch: PluginEvent[], meta: Meta<Input>) => PluginEvent[] | Promise<PluginEvent[]>
-    /** Receive a batch of non-snapshot event. */
+    /** DEPRECATED: use composeWebhook */
     exportEvents?: (events: ProcessedPluginEvent[], meta: Meta<Input>) => void | Promise<void>
-    /** Receive a single processed event. */
-    onEvent?: (event: ProcessedPluginEvent, meta: Meta<Input>) => void | Promise<void>
-    /** Ran every minute, on the minute. */
+    /** DEPRECATED: use composeWebhook */
+    onEvent?: (event: ProcessedPluginEvent | PostHogEvent, meta: Meta<Input>) => void | Promise<void>
+    /** Used for exporting a single event */
+    composeWebhook?: (event: PostHogEvent, meta: Meta<Input>) => [RequestInfo, RequestInit]
+    /** DEPRECATED:  */
     runEveryMinute?: (meta: Meta<Input>) => void
-    /** Ran every hour, on the hour. */
+    /** DEPRECATED:  */
     runEveryHour?: (meta: Meta<Input>) => void
-    /** Ran every day, on midnight. */
+    /** DEPRECATED:  */
     runEveryDay?: (meta: Meta<Input>) => void
-    /** Asynchronous jobs that can be scheduled. */
+    /** DEPRECATED */
     jobs?: {
         [K in keyof Meta<Input>['jobs']]: (
             opts: Parameters<Meta<Input>['jobs'][K]>[0],
             meta: Meta<Input>
         ) => void | Promise<void>
     }
-    /** Metrics about plugin performance that can be set by plugin devs */
+    /** DEPRECATED */
     metrics?: {
         [K in keyof Meta<Input>['metrics']]: AllowedMetricsOperations
     }
     // used for type resolution only, does not exist
     __internalMeta?: Meta<Input>
 }
 
+
+export type PluginMeta<T> = T extends { __internalMeta?: infer M } ? M : never
+
+export type Properties = Record<string, any>
+
+/** We're moving to a single PostHogEvent model use this in any future apps/plugins 
+*/
+export interface PostHogEvent {
+    /** The assigned UUIDT of the event. */
+    uuid: string
+    team_id: number
+    distinct_id: string
+    event: string
+    timestamp: string
```

Comment:
I don't know if moving to a single PostHog event model implies that we have the chance to do some radical re-naming, but if it does I would like change timestamp for something that conveys more information.

The problem is that "the timestamp of an event" can mean pretty much anything:
* When the event was sent by the client
* When the event was captured
* When the event was produced to Kafka
* When the event was ingested into CH
* And so on...

Perhaps this could be replaced for a set of timestamp type fields:
* `sent_at`
* `captured_at`
* `produced_at`
* `inserted_at`
* And so on...

Benefits include:
* Easier lag tracking of each step in the pipeline (simply subtract one field with another).
* Less ambiguity.

The big downside is of course changing every single bit of code to work with the new fields, but if this is going to be a big effort maybe we can slot that in too? If not, this should not be a blocker.

Reply:
> I don't know if moving to a single PostHog event model implies that we have the chance to do some radical re-naming, but if it does I would like change timestamp for something that conveys more information.

Yes it does.

> The problem is that "the timestamp of an event" can mean pretty much anything:

My proposal is that we move timestamp processing to happen before processEvent plugins (right after TeamResolutionStep) & this timestamp is what we'd use in CH in the end as the event `timestamp` <- the default timestamp you see in the app.

From a user perspective I'd find it simpler if there was only a single timestamp.

Reply:
We can show users a single timestamp, but I feel it will be easier to keep track of it internally if we give it a less-ambiguous name.

Reply:
I believe we should really standardise on one `timestamp`, as that's the only thing our users care about when querying events. They don't care that "it was actually ingested 8 seconds after the timestamp" or "the users clock was off by 4 minutes" or "this was sent in a batch of events where all have the same sent_at" or any of the above.

An event has a timestamp and that's it. We will need `captured_at`, `sent_at`, etc to internally make sure the `timestamp` is correct, but past that processing, those fields are not useful.

Ideally this `PostHogEvent` would already have a parsed timestamp, as it's only a string when it comes in from the users... but we then use `sent_at`, the current time, and other fields to calculate the correct timestamp.

Thus the field here will either be the one we get from the user (which will be wrong by a few seconds usually), or then we will have calculated the right timestamp... and converted it to an iso string... which we then need to convert back to a date if we want to do any processing.

Reply:
So in conclusion: We will do the processing (from sent_at etc) before any plugins in the pipeline, so this is the timestamp of the event:
```suggestion
    timestamp: Date  // in UTC timezone
```